### DISCUSSION TITLE
fp-ts: Which ring?

### DISCUSSION THREAD

**Discussion by datalek** [pagopa/technology-radar#46] [General]
Hello everyone! 👋 

Many projects[^1] at PagoPA use `fp-ts`. It's likely that, at some point, the library will be overshadowed by `effect-ts`,  with the consequent risk of not being maintained properly (or maybe not? [^2]). In your opinion, in which ring `fp-ts` should be placed?

[^1]: https://github.com/search?q=org%3Apagopa+path%3A**%2Fpackage.json+%22%5C%22fp-ts%5C%22%22+&type=code&p=1
[^2]: https://github.com/gcanti/fp-ts/discussions/1852#discussioncomment-5687771

**Comment by gunzip** [comment]
I won't put it under "adopt", since in the adopt ring we only include items when we think it would be a poor and potentially irresponsible choice _not to use them_ given the appropriate project context, which is not the case imho.

Currently, there's no signal fp-ts will be discontinued, moreover, it would be relative easy to maintain the part we use ourselves.

I wont' put it under 'hold' neither, until we have assessed `effect-ts` (read: until effect-ts goes into triel or adopt).

My vote goes to 'trial' for fp-ts and 'assess' for effect-ts.

**Reply by balanza** [reply]
Also, consider that is a package with zero dependencies and no reference to runtime-specific features.

That means that the only thing that could outdate `fp-ts` is a breaking change in TypeScript itself.

**Reply by lucacavallaro** [reply]
@balanza to "break" it yes, but JS continually evolves - we risk being reduced to using only a small subset of the language

**Reply by datalek** [reply]
I like the idea of ​​putting it on `trial`, and I think it reflects well the current state of `fp-ts` in the company. It is widely used at the moment and we are monitoring to see how it evolves.

**Comment by lucacavallaro** [comment]
In my opinion we should put `fp-ts` on `hold`
Based on our codebase, we don't appear to know functional programming and the right use of this library.

It is an utility library that we mix with imperative code (example: `if (E.isRight(something))`) without valid reason.

`effect-ts` can be introduced as `asses` if we want to see if it can be useful to us

the alternative of `fp-ts` in not necessary `effect-ts`, it can be "nothing" so to me these libraries are not related

**Reply by gunzip** [reply]
The patterns mentioned are now considered outdated, and primarily due to a lack of initial understanding. Currently, everyone has gained more confidence, and the application of these patterns has become less naive. For this reason, I don’t believe it should be considered as a factor in our decision to pause the implementation of fp-ts.

Furthermore we can introduce effect-ts as assess even without putting fp-ts on hold, which is indeed a good strategy.

**Reply by lucacavallaro** [reply]
imho we should at least try to define better the use cases of `fp-ts`, since it does not work well with infra code and external libraries (for example, is not usable with `React` without adapters)

I think that the unconditional use of `fp-ts` is one of the main problems for our productivity

**Reply by balanza** [reply]
> I think that the unconditional use of fp-ts is one of the main problems for our productivity

To be honest, I never felt held back by some `fp-ts` code. It has some drawbacks, but I still think the balance is positive. 

If we care about productivity, I think we can aim somewhere else. We used to be a high-proficient team because we invested in tools and utilities that somehow fit who we were.

Indeed we reshaped ourselves a lot, for good imho, but we didn't invest in new tools and utilities that fit who we are now. Example: repo and package fragmentation, which is a real pain in the ass.

**Reply by lucacavallaro** [reply]
IMHO the two main problems are:
1. it sometimes prevents us to use modern JavaScript features and patterns such as `AsyncIterator` (that we convert into arrays to work with this library), patterns that are extensively used on modern libraries (example: `@azure/data-tables`)
2. it requires us to write "glue code" and abstractions to work with the entire js ecosystem. This glue code it is hard to maintain and most of the time it does not add value (example: `CosmosdbModel`)

In my experience `fp-ts` can be a game changer in the `domain layer`, not in the infrastructural ones

**Reply by balanza** [reply]
Let's break the issue down a bit.

What we benefit the most is the use of `Option` and `Either` for error prevention and handling. Those data types are handy and common in modern languages (like Rust) and I advocate we keep using them even in idiomatic TypeScript. They can be used in combination with any language features.

We also benefit from `Lazy` (`TaskEither`) and `pipe` to improve composition. This practice can be used in small code chunks as well as in overall orchestrator functions: the first is still compatible with most programming styles, while the latter tends to enforce the [railway-oriented programming paradigm](https://fsharpforfunandprofit.com/rop/) which also already proved its value, but I understand is a bit of a constraint when using other programming styles. 

Lastly, we also use `sequence` and `Reader`/`Writer` for _advanced_ control flow and architecture patterns.

**Reply by balanza** [reply]
Overall, when building our engineering team at first we did benefit from having our code _standardized_ into a constrained style. We used to think that less initiative in the code organization would lead to more focus on the actual features. `fp-ts` was a cornerstone for that.

As I said, we changed a lot so it may make less sense now. We may not want standardization anymore or we may just think `fp-ts` isn't not useful for that.

I have no strong opinions, I just advocate we keep this considerations into the discussion.

**Reply by gunzip** [reply]
Indeed, the trade off of not using fp-ts would be, at the bare minimum, reverting to:

1. null checks
2. exception handling using try / catch
3. nested function calls instead of pipe 
4. some other utillity library to handle aysnchronous composition

**Reply by lucacavallaro** [reply]
I agree on usefulness of `Option` and `Either` (and maybe, there are better implementation of them in other libraries? without math abstractions and HKT emulation? we can explore this, because `fp-ts` it's hard to debug and if gli astri non sono allineati is a source of compile errors, maybe `effect-ts` can save us). 

I think that the "constrained style" has made the approach to the code much more dogmatic than practical, for example the railway oriented programming (that we use indiscriminately) despite being cool and useful in many cases, [it doesn't always fit well](https://fsharpforfunandprofit.com/posts/against-railway-oriented-programming/). But when we write backend code, just because we have to use `fp-ts` we adopt this pattern and we wrap everything (more code to maintain).

My idea is not to "stop using fp" or `fp-ts`, but make sure it is a conscious choice and regulate better the use case and the subset of the library to recommend.

**Reply by balanza** [reply]
>  make sure it is a conscious choice

💯 agree. That applies to TypeScript, too (can't count the times I had to request changes to PR due to the abuse of manual casting)

**Reply by gunzip** [reply]
![image](https://github.com/pagopa/technology-radar/assets/148224/a1ddafd8-18be-45cd-8819-59a86b8d9f7a)

**Reply by balanza** [reply]
> more dogmatic than practical

I wouldn't say dogmatic, I think sometimes you have a solution for a problem and you just want to take it for granted and move on 🤗

**Comment by lucacavallaro** [comment]
I proposed a guild to asses and study https://effect.website 

https://pagopa.atlassian.net/wiki/spaces/ENG/pages/908001327/Effect
